onnxruntime/core/optimizer/insert_cast_transformer.cc

```diff
@@ -105,14 +105,25 @@ class RemoveDuplicateCastTransformer : public GraphTransformer {
       if (node.OpType() == "Cast") {
         // if cast's next node is also cast and next cast's output type equal to cast's input type
         // remove those two cast.
+        // boolean is an exception case for this optimization
         auto src_type = node.InputDefs()[0]->Type();
         auto dst_type = node.OutputDefs()[0]->Type();
+        if (*src_type == "tensor(bool)" || *dst_type == "tensor(bool)") return Status::OK();
         auto input = node.MutableInputDefs()[0];
         int child_removed = 0;
         int num_child = 0;
+        auto output_args = graph.GetOutputs();
+        std::unordered_set<std::string> graph_outputs(output_args.size());
+        for (auto output_arg : output_args) {
+          graph_outputs.emplace(output_arg->Name());
+        }
         for (auto it = node.OutputNodesBegin(); it != node.OutputNodesEnd(); ++it) {
           const Node& output_node{*it};
           if (output_node.OpType() == "Cast") {
+            // Skip if the node's output is also the output of the graph
+            if (graph_outputs.find(output_node.OutputDefs()[0]->Name()) != graph_outputs.end()) {
```

Comment:
Could the code change to check pointer of NodeArg rather than name?

Reply:
Good points. actually we can break early.